crates/kitsune_p2p/kitsune_p2p/tests/tests.rs

```diff
@@ -2,88 +2,1148 @@ mod common;
+        .await
+        .unwrap();
+
+    let resp = tokio::time::timeout(std::time::Duration::from_secs(10), async move {
+        loop {
+            match sender
+                .rpc_single(
+                    space.clone(),
+                    agent_b.clone(),
+                    "Hello from agent a".as_bytes().to_vec(),
+                    Some(std::time::Duration::from_secs(10).as_millis() as u64),
+                )
+                .await
+            {
+                Ok(resp) => {
+                    return resp;
+                }
+                Err(_) => {
+                    tokio::time::sleep(std::time::Duration::from_millis(100)).await;
+                }
+            }
+        }
+    })
+    .await
+    .unwrap();
+
+    // Assumes that the KitsuneP2pEvent::Call handler echoes the request
+    assert_eq!("Hello from agent a".as_bytes().to_vec(), resp);
+}
+
+#[cfg(feature = "tx5")]
+#[tokio::test(flavor = "multi_thread")]
+async fn two_nodes_publish_and_fetch() {
+    holochain_trace::test_run().unwrap();
+
+    let (bootstrap_addr, _bootstrap_handle) = start_bootstrap().await;
+    let (signal_url, _signal_srv_handle) = start_signal_srv().await;
+
+    let mut harness_a = KitsuneTestHarness::try_new("host_a")
+        .await
+        .expect("Failed to setup test harness")
+        .configure_tx5_network(signal_url)
+        .use_bootstrap_server(bootstrap_addr)
+        .update_tuning_params(|mut c| {
+            // 3 seconds between gossip rounds, to keep the test fast
+            c.gossip_peer_on_success_next_gossip_delay_ms = 1000 * 3;
+            c
+        });
+
+    let sender_a = harness_a
+        .spawn()
+        .await
+        .expect("should be able to spawn node");
+
+    let mut harness_b = KitsuneTestHarness::try_new("host_b")
+        .await
+        .expect("Failed to setup test harness")
+        .configure_tx5_network(signal_url)
+        .use_bootstrap_server(bootstrap_addr)
+        .update_tuning_params(|mut c| {
+            // 3 seconds between gossip rounds, to keep the test fast
+            c.gossip_peer_on_success_next_gossip_delay_ms = 1000 * 3;
+            c
+        });
+
+    let sender_b = harness_b
+        .spawn()
+        .await
+        .expect("should be able to spawn node");
+
+    let space = Arc::new(fixt!(KitsuneSpace));
+    let agent_a = harness_a.create_agent().await;
+
+    sender_a
+        .join(space.clone(), agent_a.clone(), None, None)
+        .await
+        .unwrap();
+
+    let agent_b = harness_b.create_agent().await;
+
+    sender_b
+        .join(space.clone(), agent_b.clone(), None, None)
+        .await
+        .unwrap();
+
+    // Wait for the nodes to discover each other before publishing
+    wait_for_connected(sender_a.clone(), agent_b.clone(), space.clone()).await;
+    wait_for_connected(sender_b.clone(), agent_a.clone(), space.clone()).await;
+
+    // TODO This requires host code, does it make sense to construct valid values here?
```

Comment:
Still wondering about this, I understand the concept of a basis in mathematics and I think I follow the host implementation but undecided about trying to get something into the test harness to replicate it. For the tests I've written so far it doesn't seem to be important

Reply:
There could be bugs in the non-sharded case.. but if everything is working the basis is effectively meaningless until we start sharding > 50/100 peers, which most tests are not running at.

Reply:
Cool, this can definitely be left alone for now then